src/black-magic.js

```diff
@@ -0,0 +1,334 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from './log';
+import {IN_MUTATE_PHASE_PROP} from './dangerously-mutate';
+import {startsWith, endsWith} from './string';
+import {hasOwn} from './utils/object';
+
+
+/**
+ * A global property that checks if black magic has already been performed.
+ * @const {string}
+ */
+const INSTALLED_PROP = 'BLACK_MAGIC_INSTALLED';
+
+/**
+ * Gathers the node's ancestry tree, so that the location may be logged.
+ *
+ * @param {!Node} node
+ * @return {string}
+ */
+function ancestry(node) {
+  const tree = [node.tagName];
+  for (let c = node.parentElement; c; c = c.parentElement) {
+    const tag = c.tagName;
+    tree.push(tag);
+    if (startsWith(tag, 'AMP-')) {
+      break;
+    }
+  }
+
+  return tree.reverse().join(' > ');
+}
+
+/**
+ * Checks that this mutation is occurring in a mutation phase.
+ *
+ * @param {!Node|!Attr} nodeOrAttr
+ */
+function checkInMutationPhase(nodeOrAttr) {
+  const node = nodeOrAttr.ownerElement || nodeOrAttr;
+  const window = node.ownerDocument.defaultView;
+  if (node.isConnected !== false && !window[IN_MUTATE_PHASE_PROP]) {
+    dev().expectedError('MUTATE', 'mutation occurred outside mutation phase',
```

Comment:
I need a new error that will never be filtered.

Reply:
If you're only using this for opt-in, you can use `dev().error`.